cache_flat_mutation_reader.hh

```diff
@@ -730,20 +749,73 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
         }
     }
 
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
     if (_next_row_in_range) {
-        if (_next_row.range_tombstone_for_row() != _current_tombstone) [[unlikely]] {
-            auto tomb = _next_row.range_tombstone_for_row();
-            auto new_lower_bound = position_in_partition::before_key(_next_row.position());
-            clogger.trace("csm {}: rtc({}, {})", fmt::ptr(this), new_lower_bound, tomb);
-            push_mutation_fragment(mutation_fragment_v2(*_schema, _permit, range_tombstone_change(new_lower_bound, tomb)));
-            _lower_bound = std::move(new_lower_bound);
-            _current_tombstone = tomb;
-            _read_context.cache()._tracker.on_range_tombstone_read();
+        bool remove_row = false;
+
+        if (_read_context.tombstone_gc_state() // do not compact rows when tombstone_gc_state is not set (used in some unit tests)
+            && !_next_row.dummy()
+            && _snp->at_latest_version()
+            && _snp->at_oldest_version()) {
+
+            deletable_row& row = _next_row.latest_row();
+            tombstone range_tomb = _next_row.range_tombstone_for_row();
+            auto t = row.deleted_at();
+            t.apply(range_tomb);
+
+            auto tomb_expired = [&](row_tombstone tomb) {
+                return (t && t.max_deletion_time() < _gc_before);
+            };
+
+            auto is_row_dead = [&](const deletable_row& row) {
+                return (!row.marker().is_missing() && row.marker().is_dead(_read_time));
+            };
+
+            if (tomb_expired(t) || is_row_dead(row)) {
+                can_gc_fn always_gc = [&](tombstone) { return true; };
+                row_marker marker = row.marker();
+
+                try {
+                    with_allocator(_snp->region().allocator(), [&] {
+                        row.compact_and_expire(*_schema, range_tomb, _read_time, always_gc, _gc_before, nullptr);
+                    });
+                    remove_row = row.empty();
+                } catch (...) {
+                    row.marker() = marker;
+                    throw;
+                }
+            }
+        }
+
+        if (remove_row) {
+            partition_snapshot_row_weakref row_ref(_next_row);
```

Comment:
I think there is a mishandling of range tombstones in the case of remove_row == true.

Consider this state:

```
  (key: 0, rt: t0) ======= (key: 1, rt: t0) ========= (key: 2, rt: t1) 
``` 

range tombstones with t0 are expired, t1 is not. The row with key=1 is _next_row here and we're removing it. _current_tombstone is t0 and _lower_bound is after(key=0). After we remove the row with key=1 here, we're not advancing _lower_bound, which is an issue because it means that rt=t1 from key=2 row will extend earlier in the key range than it should, we will emit rtc to t1 with position after(key=0) instead of after(key=1). 

If we drop the row, and hence the range tombstone, we should either:
  1) emit rtc to null at the current _lower_bound (after(key=0) in this case) and clear _current_tombstone, and set _lower_bound to position_in_partition::after_key(_next_row.position()) This will reflect the fact that there is no range tombstone after it was compacted away. 
  2) keep the expired range tombstone in the output stream, so don't emit rtc and don't clear _current_tombstone, just update the _lower_bound.

Option 2 is slightly simpler to code, and I think it's fine.

We should add a unit test for this case.

Reply:
Maybe I understand something wrong but the following test works as expected when we use compaction on read as well as when we don't use compaction on read:
```
simple_schema s;
tests::reader_concurrency_semaphore_wrapper semaphore;

auto cache_mt = make_lw_shared<replica::memtable>(s.schema());

cache_tracker tracker;
row_cache cache(s.schema(), snapshot_source_from_snapshot(cache_mt->as_data_source()), tracker);

auto pk = s.make_pkey(0);
auto pr = dht::partition_range::make_singular(pk);

auto ck0 = s.make_ckey(0);
auto ck1 = s.make_ckey(1);
auto ck2 = s.make_ckey(2);

mutation m(s.schema(), pk);
s.add_row(m, ck0, "v0");
s.add_row(m, ck1, "v1");
s.add_row(m, ck2, "v2");
auto rt1 = s.make_range_tombstone(s.make_ckey_range(0, 1), gc_clock::now());
m.partition().apply_delete(*s.schema(), rt1);
cache_mt->apply(m);

tombstone_gc_state gc_state(nullptr);
auto rd1 = cache.make_reader(s.schema(), semaphore.make_permit(), query::full_partition_range, &gc_state);
auto close_rd = deferred_close(rd1);
rd1.fill_buffer().get(); // cache_flat_mutation_reader compacts cache on fill buffer        

cache_entry& entry = cache.lookup(pk);
auto& cp = entry.partition().version()->partition();

BOOST_REQUIRE(cp.find_row(*s.schema(), ck0) == nullptr);
BOOST_REQUIRE(cp.find_row(*s.schema(), ck1) == nullptr);
BOOST_REQUIRE(cp.find_row(*s.schema(), ck2) != nullptr);
```

Reply:
> tombstone_gc_state gc_state(nullptr);

Isn't tombstone compaction in cache reader disabled in this case? 

> 
> mutation m(s.schema(), pk);
> s.add_row(m, ck0, "v0");
> s.add_row(m, ck1, "v1");
> s.add_row(m, ck2, "v2");
> auto rt1 = s.make_range_tombstone(s.make_ckey_range(0, 1), gc_clock::now());
> m.partition().apply_delete(*s.schema(), rt1);
> cache_mt->apply(m);

rt1 should be expired according to the compaction state.

You also need a tombstone in the range (1, 2) which is not expired.

> BOOST_REQUIRE(cp.find_row(*s.schema(), ck0) == nullptr);
> BOOST_REQUIRE(cp.find_row(*s.schema(), ck1) == nullptr);
> BOOST_REQUIRE(cp.find_row(*s.schema(), ck2) != nullptr);

You don't verify that the next reader sees correct range tombstones so the test won't catch this bug.

Reply:
> Isn't tombstone compaction in cache reader disabled in this case?
It's enabled in this case as far as we pass gc_state to make_reader:
`auto rd1 = cache.make_reader(s.schema(), semaphore.make_permit(), query::full_partition_range, &gc_state);`

> rt1 should be expired according to the compaction state.
I tried this:
```
auto dt_noexp = gc_clock::now();
auto dt_exp = gc_clock::now() - std::chrono::seconds(s.schema()->gc_grace_seconds().count() + 1);
.....
auto rt1 = s.make_range_tombstone(s.make_ckey_range(0, 1), dt_exp);
auto rt2 = s.make_range_tombstone(s.make_ckey_range(1, 2), dt_noexp);
```
seems we wouldn't ever compact rows covered by these range tombstones because such rows considered as 'dummy' in reader and we do not compact 'dummy' rows.

Reply:
You can try to manually set the range_tombstone property on rows_entry. The situation may be hard to reach using standard apply() due to the way we represent range tombstones currently, but I think we should handle it nevertheless.

Reply:
I used the following way and seems it works as expected. Now thinking how to reproduce our situation: what checks we need to catch it
```
simple_schema s;
tests::reader_concurrency_semaphore_wrapper semaphore;

auto cache_mt = make_lw_shared<replica::memtable>(s.schema());

cache_tracker tracker;
row_cache cache(s.schema(), snapshot_source_from_snapshot(cache_mt->as_data_source()), tracker);

auto pk = s.make_pkey(0);
auto pr = dht::partition_range::make_singular(pk);

auto ck0 = s.make_ckey(0);
auto ck1 = s.make_ckey(1);
auto ck2 = s.make_ckey(2);

auto dt_noexp = gc_clock::now();
auto dt_exp = gc_clock::now() - std::chrono::seconds(s.schema()->gc_grace_seconds().count() + 1);

mutation m(s.schema(), pk);
auto rt1 = s.make_range_tombstone(s.make_ckey_range(0, 1), dt_exp);
auto rt2 = s.make_range_tombstone(s.make_ckey_range(1, 2), dt_noexp);
m.partition().apply_delete(*s.schema(), rt1);
m.partition().apply_delete(*s.schema(), rt2);
s.add_row(m, ck0, "v0");
s.add_row(m, ck1, "v1");
s.add_row(m, ck2, "v2");
cache.populate(m);

tombstone_gc_state gc_state(nullptr);

cache_entry& entry = cache.lookup(pk);
auto& cp = entry.partition().version()->partition();

BOOST_REQUIRE(cp.find_row(*s.schema(), ck0) != nullptr);
BOOST_REQUIRE(cp.find_row(*s.schema(), ck1) != nullptr);
BOOST_REQUIRE(cp.find_row(*s.schema(), ck2) != nullptr);

// make row cells to be compacted during next read
auto set_cells_timestamp_to_min = [&](deletable_row& row) {
    row.cells().for_each_cell([&] (column_id id, atomic_cell_or_collection& cell) {
        const column_definition& def = s.schema()->column_at(column_kind::clustering_key, id);

        auto cell_view = cell.as_mutable_atomic_cell(def);
        cell_view.set_timestamp(api::min_timestamp);
    });
};

set_cells_timestamp_to_min(cp.clustered_row(*s.schema(), ck0));
set_cells_timestamp_to_min(cp.clustered_row(*s.schema(), ck1));
set_cells_timestamp_to_min(cp.clustered_row(*s.schema(), ck2));

auto rd1 = cache.make_reader(s.schema(), semaphore.make_permit(), pr, &gc_state);
auto close_rd = deferred_close(rd1);
rd1.fill_buffer().get();

BOOST_REQUIRE(cp.find_row(*s.schema(), ck0) == nullptr);
BOOST_REQUIRE(cp.find_row(*s.schema(), ck1) != nullptr);
BOOST_REQUIRE(cp.find_row(*s.schema(), ck2) != nullptr);
```

Reply:
>     2. keep the expired range tombstone in the output stream, so don't emit rtc and don't clear _current_tombstone, just update the _lower_bound.
@tgrabiec Tomasz, we should shift _lower_bound to the next position, right? I did that but see the following:
the test case is:
```
mutation m(s.schema(), pk);
auto rt1 = s.make_range_tombstone(s.make_ckey_range(2, 3), dt_noexp);
auto rt2 = s.make_range_tombstone(s.make_ckey_range(1, 2), dt_exp);
m.partition().apply_delete(*s.schema(), rt1);
m.partition().apply_delete(*s.schema(), rt2);
s.add_row(m, ck0, "v0");
s.add_row(m, ck1, "v1");
s.add_row(m, ck2, "v2");
s.add_row(m, ck3, "v3");
```
we compact rows with keys 1 and 2 during first read
but on second read I see the trace:
```
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: move_to_next_entry(), curr={position: clustered, ckp{000c636b30303030303030303030}, 0}
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: next={position: clustered, ckp{000c636b30303030303030303031}, -1}, cont=true, in_range=true
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: reading_from_cache, range=[{position: clustered, ckp{000c636b30303030303030303030}, 1}, {position: clustered, ckp{}, 1}), next={position: clustered, ckp{000c636b30303030303030303031}, -1}, valid=true, rt={tombstone: none}
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: next={cursor: position={position: clustered, ckp{000c636b30303030303030303031}, -1}, cont=true, rt={tombstone: none}}, snp=0x601000160bd0, current=[{v=0, pos={position: clustered, ckp{000c636b30303030303030303031}, -1}, cont=true, rt={tombstone: none}, row_rt={tombstone: none}}], heap=[
  ], latest_iterator=[{position: clustered, ckp{000c636b30303030303030303031}, -1}]}
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: copy_from_cache, next_row_in_range=true, next={cursor: position={position: clustered, ckp{000c636b30303030303030303031}, -1}, cont=true, rt={tombstone: none}}, snp=0x601000160bd0, current=[{v=0, pos={position: clustered, ckp{000c636b30303030303030303031}, -1}, cont=true, rt={tombstone: none}, row_rt={tombstone: none}}], heap=[
  ], latest_iterator=[{position: clustered, ckp{000c636b30303030303030303031}, -1}]}
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: move_to_next_entry(), curr={position: clustered, ckp{000c636b30303030303030303031}, -1}
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: next={position: clustered, ckp{000c636b30303030303030303032}, 1}, cont=false, in_range=true
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: start_reading_from_underlying(), range=[{position: clustered, ckp{000c636b30303030303030303031}, -1}, {position: clustered, ckp{000c636b30303030303030303032}, 1})
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: partition does not exist
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: underlying done, in_range=true, same=true, next={cursor: position={position: clustered, ckp{000c636b30303030303030303032}, 1}, cont=false, rt={tombstone: none}}, row_rt={tombstone: timestamp=-9223372036854775806, deletion_time=1683377113}, snp=0x601000160bd0, current=[{v=0, pos={position: clustered, ckp{000c636b30303030303030303032}, 1}, cont=false, rt={tombstone: timestamp=-9223372036854775806, deletion_time=1683377113}, row_rt={tombstone: timestamp=-9223372036854775806, deletion_time=1683377113}}], heap=[
  ], latest_iterator=[{position: clustered, ckp{000c636b30303030303030303032}, 1}]}
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: set_continuous({position: clustered, ckp{000c636b30303030303030303032}, 1}), rt={tombstone: none}
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: dropping unnecessary dummy at {position: clustered, ckp{000c636b30303030303030303031}, -1}
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: reading_from_cache, range=[{position: clustered, ckp{000c636b30303030303030303032}, 1}, {position: clustered, ckp{}, 1}), next={position: clustered, ckp{000c636b30303030303030303032}, 1}, valid=true, rt={tombstone: none}
INFO  2023-05-16 15:45:14,352 [shard 0] cache - csm 0x60100022f300: next={cursor: position={position: clustered, ckp{000c636b30303030303030303032}, 1}, cont=false, rt={tombstone: none}}, row_rt={tombstone: timestamp=-9223372036854775806, deletion_time=1683377113}, snp=0x601000160bd0, current=[{v=0, pos={position: clustered, ckp{000c636b30303030303030303032}, 1}, cont=false, rt={tombstone: timestamp=-9223372036854775806, deletion_time=1683377113}, row_rt={tombstone: none}}], heap=[
  ], latest_iterator=[{position: clustered, ckp{000c636b30303030303030303032}, 1}]}
```
seems we still have rtc's in the cache and trying to touch compacted ckeys? Im afraid that's not the result we wanna get?
on third read everyting works as expected: reader reads non-compacted rows only.
Changing _lower_bound doesn't affect situation at all. Maybe we should compact expired rtc's as well as clustering rows?